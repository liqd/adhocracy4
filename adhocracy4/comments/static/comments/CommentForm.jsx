var config = require('../../../static/config')
var Alert = require('../../../static/Alert')

var React = require('react')
var PropTypes = require('prop-types')
var django = require('django')

class CommentForm extends React.Component {
  constructor (props) {
    super(props)

    this.state = {comment: ''}
  }

  handleTextChange (e) {
    this.setState({comment: e.target.value})
  }

  handleSubmit (e) {
    e.preventDefault()
    var comment = this.state.comment.trim()
    if (!comment) {
      return
    }
    this.props.onCommentSubmit({
      comment: comment,
      urlReplaces: {
        objectPk: this.props.subjectId,
        contentTypeId: this.props.subjectType
      }
    }, this.props.parentIndex).then(() => {
      this.setState({comment: ''})
    })
  }

  render () {
    if (this.context.isAuthenticated && !this.props.isReadOnly) {
      return (
        <form className="general-form" onSubmit={this.handleSubmit.bind(this)}>
          {this.props.error &&
            <Alert type="danger" message={django.gettext('Error while submitting your comment!')} onClick={this.props.handleErrorClick} />
          }
          <div className="form-group">
            <textarea
              autoFocus={this.props.grabFocus}
              rows={this.props.rows}
              className="form-control"
              placeholder={django.gettext('Your comment here')}
              onChange={this.handleTextChange.bind(this)} required="required" value={this.state.comment} />
          </div>
          <input type="submit" value={django.gettext('post')} className="submit-button" />
        </form>
      )
    } else if (!this.props.isReadOnly) {
      return (
        <div className="comments_login">
          <a href={config.getLoginUrl()}>{django.gettext('Please login to comment')}</a>
        </div>
      )
    } else {
      return (
        <div>
          {django.gettext('The currently active phase doesn\'t allow to comment.')}
        </div>
      )
    }
  }
}

CommentForm.contextTypes = {
  isAuthenticated: PropTypes.bool
}

module.exports = CommentForm
