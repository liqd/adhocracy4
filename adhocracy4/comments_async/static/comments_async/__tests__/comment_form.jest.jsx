import React from 'react'
import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import CommentForm from '../comment_form'

describe('CommentForm Component', () => {
  const defaultProps = {
    agreedTermsOfUse: false,
    useTermsOfUse: false,
    subjectId: 108,
    subjectType: 107,
    hasCommentingPermission: true,
    onCommentSubmit: jest.fn().mockResolvedValue()
  }

  test('renders CommentForm', () => {
    const tree = render(<CommentForm {...defaultProps} />)
    expect(tree).toMatchSnapshot()
  })

  test('submit button cannot be pressed multiple times', async () => {
    render(<CommentForm {...defaultProps} />)

    const textarea = screen.getByLabelText(/Your comment/)
    const submitButton = screen.getByRole('button', { name: 'Post' })
    // button should be enabled before submit
    expect(submitButton).not.toBeDisabled()

    fireEvent.change(textarea, { target: { value: 'this is a new comment' } })
    expect(screen.getByText(/this is a new comment/)).toBeInTheDocument()
    // click button first time
    fireEvent.click(submitButton)
    // button should be disabled
    expect(submitButton).toBeDisabled()
    // click again
    fireEvent.click(submitButton)
    fireEvent.click(submitButton)

    await waitFor(() => {
      // button should be enabled again after submit
      expect(submitButton).not.toBeDisabled()
      // check that form submit has only been triggered once
      expect(defaultProps.onCommentSubmit).toHaveBeenCalledTimes(1)
    })
  })
})
