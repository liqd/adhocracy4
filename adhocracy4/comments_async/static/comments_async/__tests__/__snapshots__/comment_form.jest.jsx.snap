// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentForm Component renders CommentForm 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="a4-comments__comment-form__heading-comments-allowed"
      >
        Join the discussion
      </h3>
      <form
        class="a4-comments__comment-form__form"
        id="id-comment-formundefined"
      >
        <div
          class="form-group"
          id="group_id_textarea-topundefined"
        >
          <label
            for="id_textarea-topundefined"
          >
            Your comment
          </label>
          <textarea
            aria-describedby="id_char-countundefined"
            aria-invalid="false"
            class="a4-comments__textarea--small"
            id="id_textarea-topundefined"
            name="id_textarea-topundefined"
            style="height: 75px;"
          />
          <p
            aria-live="polite"
            class="a4-comments__char-count"
            id="id_char-countundefined"
          >
            <span
              aria-hidden="true"
            >
              0
              /4000
              <span
                class="a4-comments__char-count-word"
              >
                 characters
              </span>
            </span>
            <span
              class="a4-sr-only"
            >
              0
              /4000
               characters
            </span>
          </p>
        </div>
        <div
          class="row a4-comments__comment-form__terms-and-buttons"
        >
          <div
            class="a4-comments__comment-form__terms-of-use"
          />
          <div
            class="a4-comments__comment-form__actions"
          >
            <div
              class="a4-comments__comment-form__actions__left"
            />
            <div
              class="a4-comments__comment-form__actions__right"
            >
              <button
                class="btn btn--default a4-comments__submit-input"
                type="submit"
              >
                Post
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <h3
      class="a4-comments__comment-form__heading-comments-allowed"
    >
      Join the discussion
    </h3>
    <form
      class="a4-comments__comment-form__form"
      id="id-comment-formundefined"
    >
      <div
        class="form-group"
        id="group_id_textarea-topundefined"
      >
        <label
          for="id_textarea-topundefined"
        >
          Your comment
        </label>
        <textarea
          aria-describedby="id_char-countundefined"
          aria-invalid="false"
          class="a4-comments__textarea--small"
          id="id_textarea-topundefined"
          name="id_textarea-topundefined"
          style="height: 75px;"
        />
        <p
          aria-live="polite"
          class="a4-comments__char-count"
          id="id_char-countundefined"
        >
          <span
            aria-hidden="true"
          >
            0
            /4000
            <span
              class="a4-comments__char-count-word"
            >
               characters
            </span>
          </span>
          <span
            class="a4-sr-only"
          >
            0
            /4000
             characters
          </span>
        </p>
      </div>
      <div
        class="row a4-comments__comment-form__terms-and-buttons"
      >
        <div
          class="a4-comments__comment-form__terms-of-use"
        />
        <div
          class="a4-comments__comment-form__actions"
        >
          <div
            class="a4-comments__comment-form__actions__left"
          />
          <div
            class="a4-comments__comment-form__actions__right"
          >
            <button
              class="btn btn--default a4-comments__submit-input"
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
